File: modules/module-01-core-experience/exercises/orderService.js
/**
 * Order Service - Enterprise JavaScript Module
 *
 * Exercise: Use GitHub Copilot inline suggestions to complete
 * each function. Write a comment describing what you want,
 * then let Copilot suggest the implementation.
 *
 * Tips:
 * - Press Tab to accept a suggestion
 * - Press ⌘→ (Mac) / Ctrl→ (Win) to accept word-by-word
 * - Press ⌥] / Alt] to see alternative suggestions
 * - Press Esc to dismiss
 */

class OrderService {
  constructor(orders = []) {
    this.orders = orders;
  }

  /**
   * Calculate the total amount for a single order
   * by summing up item.price * item.quantity for all items.
   * Apply a discount percentage if provided (e.g., 10 means 10% off).
   * Return 0 if the order has no items.
   *
   * @param {Object} order - The order object with items array
   * @param {number} discountPercent - Optional discount percentage (0-100)
   * @returns {number} The total amount after discount
   */
  calculateOrderTotal(order, discountPercent = 0) {
    // TODO: Use Copilot to complete this method
  }

  /**
   * Filter orders by status.
   * Valid statuses: 'pending', 'processing', 'shipped', 'delivered', 'cancelled'
   * Throw an Error if an invalid status is provided.
   * Return an empty array if no orders match.
   *
   * @param {string} status - The status to filter by
   * @returns {Array} Filtered orders
   */
  filterOrdersByStatus(status) {
    // TODO: Use Copilot to complete this method
  }

  /**
   * Get orders within a date range (inclusive).
   * Compare using order.createdAt (a Date object).
   * Validate that startDate is before endDate; throw Error if not.
   * Return orders sorted by createdAt ascending.
   *
   * @param {Date} startDate - Start of the date range
   * @param {Date} endDate - End of the date range
   * @returns {Array} Orders within the date range, sorted ascending
   */
  getOrdersByDateRange(startDate, endDate) {
    // TODO: Use Copilot to complete this method
  }

  /**
   * Calculate the average order value across all orders.
   * Use calculateOrderTotal for each order (no discount).
   * Return 0 if there are no orders.
   * Round to 2 decimal places.
   *
   * @returns {number} The average order value
   */
  getAverageOrderValue() {
    // TODO: Use Copilot to complete this method
  }

  /**
   * Group orders by customer ID.
   * Return an object where keys are customerIds and values are arrays of orders.
   * Sort each customer's orders by createdAt descending (newest first).
   *
   * @returns {Object} Orders grouped by customer ID
   */
  groupOrdersByCustomer() {
    // TODO: Use Copilot to complete this method
  }

  /**
   * Find the top N customers by total spending.
   * Return an array of objects: { customerId, totalSpent, orderCount }
   * Sorted by totalSpent descending.
   * If n is greater than the number of customers, return all customers.
   *
   * @param {number} n - Number of top customers to return
   * @returns {Array} Top N customers by spending
   */
  getTopCustomers(n = 5) {
    // TODO: Use Copilot to complete this method
  }
}

module.exports = OrderService;

File: modules/module-01-core-experience/exercises/OrderService.java
package com.enterprise.demo;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Order Service - Enterprise Java Module
 *
 * Exercise: Use GitHub Copilot inline suggestions to complete
 * each method. Write a comment describing what you want,
 * then let Copilot suggest the implementation.
 *
 * Tips:
 * - Press Tab to accept a suggestion
 * - Press ⌘→ (Mac) / Ctrl→ (Win) to accept word-by-word
 * - Press ⌥] / Alt] to see alternative suggestions
 * - Press Esc to dismiss
 */
public class OrderService {

    public enum OrderStatus {
        PENDING, PROCESSING, SHIPPED, DELIVERED, CANCELLED
    }

    public record OrderItem(String productId, String productName, BigDecimal price, int quantity) {}

    public record Order(
        String orderId,
        String customerId,
        List<OrderItem> items,
        OrderStatus status,
        LocalDate createdAt
    ) {}

    public record CustomerSummary(String customerId, BigDecimal totalSpent, int orderCount) {}

    private final List<Order> orders;

    public OrderService(List<Order> orders) {
        this.orders = orders != null ? new ArrayList<>(orders) : new ArrayList<>();
    }

    /**
     * Calculate the total amount for a single order
     * by summing up item.price() * item.quantity() for all items.
     * Apply a discount percentage if provided (e.g., 10 means 10% off).
     * Return BigDecimal.ZERO if the order has no items or order is null.
     * Use BigDecimal arithmetic with RoundingMode.HALF_UP and scale of 2.
     *
     * @param order           the order to calculate
     * @param discountPercent the discount percentage (0-100), 0 means no discount
     * @return the total amount after discount
     */
    public BigDecimal calculateOrderTotal(Order order, int discountPercent) {
        // TODO: Use Copilot to complete this method
    }

    /**
     * Filter orders by status.
     * Throw IllegalArgumentException if status is null.
     * Return an empty list if no orders match.
     *
     * @param status the status to filter by
     * @return filtered list of orders
     */
    public List<Order> filterOrdersByStatus(OrderStatus status) {
        // TODO: Use Copilot to complete this method
    }

    /**
     * Get orders within a date range (inclusive on both ends).
     * Compare using order.createdAt().
     * Validate that startDate is before or equal to endDate; throw
     * IllegalArgumentException if not, or if either date is null.
     * Return orders sorted by createdAt ascending.
     *
     * @param startDate start of the date range (inclusive)
     * @param endDate   end of the date range (inclusive)
     * @return orders within the range, sorted ascending by date
     */
    public List<Order> getOrdersByDateRange(LocalDate startDate, LocalDate endDate) {
        // TODO: Use Copilot to complete this method
    }

    /**
     * Calculate the average order value across all orders.
     * Use calculateOrderTotal for each order with 0 discount.
     * Return BigDecimal.ZERO if there are no orders.
     * Round to 2 decimal places with RoundingMode.HALF_UP.
     *
     * @return the average order value
     */
    public BigDecimal getAverageOrderValue() {
        // TODO: Use Copilot to complete this method
    }

    /**
     * Group orders by customer ID.
     * Return a Map where keys are customerIds and values are lists of orders.
     * Sort each customer's orders by createdAt descending (newest first).
     *
     * @return orders grouped by customer ID
     */
    public Map<String, List<Order>> groupOrdersByCustomer() {
        // TODO: Use Copilot to complete this method
    }

    /**
     * Find the top N customers by total spending.
     * Return a list of CustomerSummary records sorted by totalSpent descending.
     * If n is greater than the number of customers, return all customers.
     * Throw IllegalArgumentException if n is less than 1.
     *
     * @param n number of top customers to return
     * @return list of top N customers by spending
     */
    public List<CustomerSummary> getTopCustomers(int n) {
        // TODO: Use Copilot to complete this method
    }
}
